src/v/kafka/security/authorizer.h

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "kafka/security/acl.h"
+#include "kafka/security/acl_store.h"
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "seastarx.h"
+
+#include <seastar/core/sstring.hh>
+#include <seastar/util/bool_class.hh>
+
+#include <absl/container/flat_hash_set.h>
+#include <fmt/core.h>
+
+namespace kafka {
+
+/*
+ * Primary interface for request authorization and management of ACLs.
+ */
+class authorizer final {
+public:
+    // allow operation when no ACL is found
+    using allow_if_empty = ss::bool_class<struct allow_if_empty_type>;
+
+    authorizer()
+      : authorizer(allow_if_empty::no) {}
+
+    explicit authorizer(allow_if_empty allow)
+      : _allow_if_empty(allow) {}
+
+    /*
+     * Add ACL bindings to the authorizer.
+     */
+    void add_bindings(const std::vector<acl_binding>& bindings) {
+        _store.add_bindings(bindings);
+    }
+
+    /*
+     * Remove ACL bindings that match the filter(s).
+     */
+    void remove_bindings(const std::vector<acl_binding_filter>& filters) {
+        _store.remove_bindings(filters);
+    }
+
+    /*
+     * Retrieve ACL bindings that match the filter.
+     */
+    std::vector<acl_binding> acls(const acl_binding_filter& filter) const {
+        return _store.acls(filter);
+    }
+
+    /*
+     * Authorize an operation on a resource. The type of resource is deduced by
+     * the type `T` of the name of the resouce (e.g. `model::topic`).
+     */
+    template<typename T>
+    bool authorized(
+      const T& resource_name,
+      acl_operation operation,
+      const acl_principal& principal,
+      const acl_host& host) const {
+        auto type = get_resource_type<T>();
+        auto acls = _store.find(type, resource_name());
+
+        if (_superusers.contains(principal.string())) {
+            return true;
+        }
+
+        if (acls.empty()) {
+            return bool(_allow_if_empty);
+        }
+
+        // check for deny
+        if (acls.contains(operation, principal, host, acl_permission::deny)) {
+            return false;
+        }
+
+        // check for allow
+        auto ops = acl_implied_ops(operation);
+        return std::any_of(
+          ops.cbegin(),
+          ops.cend(),
+          [&acls, &principal, &host](acl_operation operation) {
+              return acls.contains(
+                operation, principal, host, acl_permission::allow);
+          });
+    }
+
+    void add_superuser(ss::sstring principal) {
+        _superusers.emplace(std::move(principal));
+    }
+
+private:
+    acl_store _store;
+    absl::flat_hash_set<ss::sstring> _superusers;
```

Comment:
a note on what `_superusers` are would be helpful and how they relate to _store.

Reply:
fixed